Fix data type

ylipacbio opened the pull request:
Fixed data type in many places.
Added -Wall -Wextra to pbdata/makefile, hdf/makefile, alignment/makefile
Removed all warnings.
Included pull request 109 in commits e2c284b, 8648430, 8469992, d651ba9, 7b9e7b8

ylipacbio commented:
This pull request contains lots of small commits which require extra code review.  @nlhepler @cdunn2001 @pbjd @vrainish-pacbio 

An upcoming pull request will be created on the blasr side to work with this commit.

ylipacbio commented:
Chris, this commit should contain (be on top of) pull request 107 (commit 7609d4b). But 107 hasn't been merged to master. what should we do?

pb-cdunn commented:
> Chris, this commit should contain (be on top of) pull request 107 (commit 7609d4b). But 107 hasn't been merged to master. what should we do?

Best is to merge first. 107 looks fine. I'd merge it right away.

However, if you really just want a code-review which excludes 107, before 107 is merged, then you need 107 on another branch. You can create a phony pull-request against that branch. You could even treat that as real, and after 107 is merged, that new PR will be mergeable immediately.

pb-cdunn commented on pbdata/utilsImpl.hpp:
```diff
@@ -24,7 +24,7 @@ T_Int CeilOfFraction(T_Int num, T_Int denom) {
 
 template<typename T>
-inline T* ProtectedNew(unsigned long size) {
+inline T* ProtectedNew(uint64_t size) {
```
This doesn't really make sense. On a 32-bit machine, you cannot allocate a 64-bit array. This *should* be `size_t`, which is why `size_t` is used in [standard library routines](http://www.cplusplus.com/reference/cstdlib/malloc/).

pb-cdunn commented on pbdata/sam/ReferenceSequence.hpp:
```diff
@@ -21,7 +21,7 @@ class SAMReferenceSequence {
     return "";
   }
-  unsigned int GetLength() {
+  uint64_t GetLength() {
```
Basically, lengths should be `size_t` since a 32-bit machine will have a tighter limit.

64-bit integers are for 64-bit math. If you don't have any 64-bit math, then you don't need `uint64_t` anywhere! `size_t` is better because it allows 32-bit programs to use less memory.

Soon, this knowledge will be obsolete, since everything will be 64-bit architecture, but I'm just sayin'.

pb-cdunn replied:
Well, if you're just trying to avoid bugs, 64-bit types are fine for consistency. 32-bit users can grumble.

In that case, I don't see any particular problems with this PR.

pb-cdunn commented on pbdata/reads/ScanData.cpp:
```diff
@@ -37,26 +37,29 @@
     if (not baseMap.empty()) {
         baseMapStr = "    ";
         for (auto it = baseMap.begin(); it != baseMap.end(); ++it){
-            if (it->second > 4 or it->second < 0) {
```
Why remove this test? Expensive? It certainly can prevent a seg-fault.

ylipacbio replied:
Fixed

pb-cdunn commented on pbdata/reads/ScanData.cpp:
```diff
@@ -37,26 +37,29 @@
 
 std::string ScanData::BaseMapToStr(const std::map<char, size_t> & baseMap) {
     std::string baseMapStr = ""; //4 dye channels.
     if (not baseMap.empty()) {
```
I don't think this `if` is ever needed.

ylipacbio replied:
Fixed

pb-cdunn commented on pbdata/reads/PulseFile.cpp:
```diff
@@ -98,15 +97,20 @@ void PulseFile::CopyReadAt(uint32_t plsReadIndex, int *baseToPulseIndex, SMRTSeq
     }
     if (startFrame.size() > 0) {
-        Realloc(read.startFrame, read.length);
+        OK = OK and Realloc(read.startFrame, read.length);
```
If `OK` is false, `Realloc()` will not happen. But `StoreField()` is called anyway. Is that correct?

ylipacbio replied:
En... there is a problem here. We should not copy anything if Realloc fails. 
Fixed

pb-cdunn commented on hdf/makefile:
```diff
@@ -4,8 +4,9 @@ THISDIR:=$(dir $(realpath $(lastword $(MAKEFILE_LIST))))
-INCLUDES += ${LIBPBDATA_INC} ${HDF5_INC} ${PBBAM_INC} ${HTSLIB_INC} ${BOOST_INC}
+CXXOPTS  += -std=c++11 -pedantic -Wall -Wextra
+INCLUDES += ${LIBPBDATA_INC}
+SYSINCLUDES = ${HDF5_INC} ${PBBAM_INC} ${HTSLIB_INC} ${BOOST_INC}
```
I think this distinction matters only when producing `.d` files, which are not really helpful anyway. I'd drop the `.d` files if nobody really needs them.

ylipacbio replied:
I am not sure if .d files help with make -j 8 ? Currently I am happy that blasr_libcpp can be compiled with -j 8, and I don't want to lose it.
Delete *.d if it does not have any affect on -j 8.

pb-cdunn commented on hdf/makefile:
```diff
@@ -4,8 +4,9 @@ THISDIR:=$(dir $(realpath $(lastword $(MAKEFILE_LIST))))
 
-CXXOPTS  += -std=c++11 -pedantic
-INCLUDES += ${LIBPBDATA_INC} ${HDF5_INC} ${PBBAM_INC} ${HTSLIB_INC} ${BOOST_INC}
+CXXOPTS  += -std=c++11 -pedantic -Wall -Wextra
```
`-Wonderful`!

pb-cdunn commented on hdf/HDFPulseDataFile.cpp:
```diff
@@ -34,15 +34,18 @@ HDFPulseDataFile::HDFPulseDataFile() {
+    GetAllReadLengths(offset_);
+    // type of read length of a single read : DNALength
+    // type of total read length of all reads in plx.h5: DSLength
+    eventOffset = std::vector<DSLength>(offset_.begin(), offset_.end());
```
I think you could use [`vector::assign()`](www.cplusplus.com/reference/vector/vector/assign/) here, to avoid an extra temporary array on the heap and extra copying.

pb-cdunn commented on hdf/HDFPulseCallsWriter.cpp:
```diff
@@ -285,6 +285,7 @@ bool HDFPulseCallsWriter::_CheckRead(const PacBio::BAM::BamRecord & read,
                                      const uint32_t qvLength,
                                      const std::string & qvName) {
     // Tag 'pb' will not be available until 3.0.1, bug 29486
+    (void)(read); (void)(qvLength); (void)(qvName);
```
Interesting. TIL.

pb-cdunn commented on alignment/datastructures/alignment/FilterCriteria.cpp:
```diff
@@ -140,6 +140,10 @@ HitPolicy::Apply(const std::vector<T_AlignmentCandidate*> alnPtrs,
     std::vector<T_AlignmentCandidate*> ret = alnPtrs;
     // Sort AlignmentCandidate pointers according to score and target start position.
     ScoreSign sign = this->Sign();
```
ScoreSign const sign = this->Sign();

That way, if you accidentally type

    if (sign = x) {

the compiler would catch your mistake!

pb-cdunn commented:
That's all I found. LGTM!

ylipacbio commented:
Chris, I fixed issues that you have mentioned, merged 107, rebased to origin master and push -f the latest code. 

Would you please take a look and see if it looks fine? Thanks!

pb-cdunn commented:
@ylipacbio, We should merge this right away if it's working. Please don't wait just because of the new warnings. Those can be fixed in a *new* pull-request. Since the PR was so large to begin with, you cannot just fix something and ask someone to look at it again.

I really, really, really want to see smaller, more frequent pull-request.

pb-cdunn commented:
```
